src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs

```diff
@@ -570,61 +612,95 @@ private async Task ProcessSubscriptionAsync()
-                                    await SendHeartBeat("Didn't find any documents to send and more then 1000ms passed");
+                                    updateBatchPerformanceStats();
+                                    
+                                    if (sendingCurrentBatchStopwatch.ElapsedMilliseconds > 1000)
+                                        await SendHeartBeat("Didn't find any documents to send and more then 1000ms passed");
+                                    
+                                    using (docsContext.OpenReadTransaction())
+                                    {
+                                        var globalEtag = useRevisions
+                                            ? TcpConnection.DocumentDatabase.DocumentsStorage.RevisionsStorage.GetLastRevisionEtag(docsContext, Subscription.Collection)
+                                            : TcpConnection.DocumentDatabase.DocumentsStorage.GetLastDocumentEtag(docsContext.Transaction.InnerTransaction, Subscription.Collection);
 
-                                using (docsContext.OpenReadTransaction())
-                                {
-                                    var globalEtag = useRevisions ?
-                                        TcpConnection.DocumentDatabase.DocumentsStorage.RevisionsStorage.GetLastRevisionEtag(docsContext, Subscription.Collection) :
-                                        TcpConnection.DocumentDatabase.DocumentsStorage.GetLastDocumentEtag(docsContext.Transaction.InnerTransaction, Subscription.Collection);
+                                        if (globalEtag > _startEtag)
+                                            continue;
+                                    }
 
-                                    if (globalEtag > _startEtag)
+                                    AssertCloseWhenNoDocsLeft();
+                                    
+                                    if (await WaitForChangedDocuments(replyFromClientTask)) 
                                         continue;
                                 }
+                            }
 
-                                AssertCloseWhenNoDocsLeft();
-
-                                if (await WaitForChangedDocuments(replyFromClientTask))
-                                    continue;
+                            // using (batchParentScope.For(SubscriptionBatchOperations.AckScope))
+                            using (batchParentScope.For(SubscriptionOperationScope.Ack))
+                            {
+                                batchParentScope.RecordStartWaitingForClientAck(DateTime.UtcNow);
+                                
+                                (replyFromClientTask, subscriptionChangeVectorBeforeCurrentBatch) =
+                                    await WaitForClientAck(replyFromClientTask, subscriptionChangeVectorBeforeCurrentBatch);
+
+                                batchParentScope.RecordClientAckTime(Stats.LastAckReceivedAt);
```

Comment:
I admit those dates seems to be suspicious for me. what we don't use current time here?

Reply:
`Stats.LastAckReceivedAt `is the actual time when the **ack** was received
It is set inside method `WaitForClientAck`